src/pages/learn/performance.mdx

```diff
@@ -0,0 +1,77 @@
+# Performance
+
+<p className="learn-subtitle">Optimize the execution and delivery of GraphQL responses</p>
+
+At first glance, GraphQL requests may seem challenging to cache given that the API is served through a single endpoint and you may not know in advance what fields a client will include in an operation.
+
+In practice, however, GraphQL is as cacheable as any API that enables parameterized requests, such as a REST API that allows clients to specify different query parameters for a particular endpoint. There are many ways to optimize GraphQL requests on the client and server sides, as well as in the transport layer, and different GraphQL server and client libraries will often have common caching features built directly into them.
+
+On this page, we'll explore several different tactics that can be leveraged in GraphQL clients and servers to optimize how data is fetched from the API.
+
+## `GET` requests for queries
+
+GraphQL implementations that adhere to the [GraphQL over HTTP specification](https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md) will support the `POST` HTTP method by default, but may also support `GET` requests for query operations.
+
+Using `GET` can improve query performance because requests made with this HTTP method are typically considered cacheable by default and can help facilitate HTTP caching or the use of a content delivery network (CDN) when caching-related headers are provided in the server response. 
+
+However, because browsers and CDNs impose size limits on URLs, it may not be possible to send a large document for complex operations in the query string of the URL. Using _persisted queries_, either in the form of _trusted documents_ or _automatic persisted queries_, will allow the client to send a hash of the query instead, and the server can look up the full version of the document by looking up the hash in a server-side store before validating and executing the operation.
+
+Sending hashed queries instead of their plaintext versions has the additional benefit of reducing the amount of data sent by the client in the network request.
+
+## The N+1 Problem
+
+GraphQL is designed in a way that allows you to write clean code on the server, where every field on every type has a focused single-purpose function for resolving that value. However, without additional consideration, a naive GraphQL service could be very "chatty" or repeatedly load data from your databases.
+
+Consider the following query—to fetch a hero along with a list of their friends, we can imagine that as the field resolvers execute there will be one request to the underlying data source to get the character object, and then three subsequent requests to load the friends' character objects:
+
+```graphql
+# { "graphiql": true }
+query HeroWithFriends {
+  # 1 request for the hero
+  hero {
+    name
+    # 3 more requests for each friend
+    friends {
+      name
+    }
+  }
+}
+```
```

Comment:
The idea here was a more naive implementation GraphQL wraps something like a REST API) and you're making 3 separate requests for each friend character to a `/character/{id}` endpoint to get them by their IDs. Perhaps I can add more context to the example rather than building out more types and fields in schema to illustrate this?

Reply:
In a REST API I'd expect this to be solved via a single request to something like `/users/:id/friends` - I think you might be suggesting that a request to `/users/:id` would return `{friend_ids: [...]}` and then you'd make individual requests to `/users/:id` for each friend, but I don't commonly see this in practice.

If we substitute "friends" for "repositories", then GitHub's REST API [works in the way that I'd suggest above](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repositories-for-a-user), i.e. a `GET` to `/users/{username}/repos`.

So to update the previous resolvers to use REST instead, you could imagine something like:

```ts
function Query_hero(_, __, { userId }) {
  return rest`/users/${userId}`;
}
function Hero_friends(hero) {
  return rest`/users/${hero.id}/friends`;
}
```

Again, there's no N+1 problem here. I understand the desire to not expand the schema further, but I don't think your example clearly demonstrates the N+1 problem even for REST APIs.

Reply:
The critical thing is that in your example the only resolver that is called more than once is the `name` resolver, and that is likely a trivial resolver. We need to select a non-trivial field _inside_ of a list field. So you can demonstrate this with the existing schema using something like:

````suggestion
```graphql
# { "graphiql": true }
query HeroWithFriends {
  # 1 request for the hero
  hero {
    name
    # 1 request for the hero's N friends
    friends {
      name
      # N planet requests - one for each of the hero's N friends
      planet {
        name
        climate
      }
    }
  }
}
```
````

Here you might have a resolver like:

```ts
function Hero_planet(hero) {
  return GET`/planets/${hero.planet_id}`
  // Or: return sql`select * from planets where id = ${hero.planet_id}`
}
```

This resolver suffers the N+1 problem because it will be called N times (once for each hero in the parent selection set).

It's not ideal that this has the extra layer, but it does avoid having to change the schema.

Reply:
OK, I just pushed https://github.com/graphql/graphql.github.io/pull/1818/commits/19cedbd03820ca356c5308757ac783c33c3da3cf with a modified N+1. I used the `starships` field instead so we could work with the existing schema.